lettuce/gencuda/ModuleGenerator.py

```diff
@@ -0,0 +1,205 @@
+import os
+import shutil
+
+from lettuce.gencuda import *
+from lettuce.gencuda.util import pretty_print_c_, pretty_print_py_
+
+py_frame = '''
+import torch
+import os
+
+import lettuce.simulation
+
+# on windows add cuda path for
+# native module to find all dll's
+if os.name == 'nt':
+    os.add_dll_directory(os.path.join(os.environ['CUDA_PATH'], 'bin'))
+
+import {module}
+
+{py_buffer}
+
+def resolve(stencil: str, equilibrium: str, collision: str, stream: str = None,
+            support_no_stream: bool = None, support_no_collision: bool = None):
+    """"""
+
+    stream = stream if stream is not None else "standard"
+
+    extra = ''
+    if support_no_stream is not None and support_no_stream:
+        extra = '_nsm'
+    if support_no_collision is not None and support_no_collision:
+        extra = '_ncm'
+
+    return globals()[f"{{stencil}}_{{equilibrium}}_{{collision}}_{{stream}}{{extra}}"]
+'''
+
+pybind_frame = '''
+#if _MSC_VER && !__INTEL_COMPILER
+#pragma warning ( push )
+#pragma warning ( disable : 4067 )
+#pragma warning ( disable : 4624 )
+#endif
+
+#include <torch/extension.h>
+
+#if _MSC_VER && !__INTEL_COMPILER
+#pragma warning ( pop )
+#endif
+
+{includes}
+
+PYBIND11_MODULE(TORCH_EXTENSION_NAME, m)
+{{
+    {definitions}
+}}
+'''
+
+
+class ModuleMatrix:
+    """
+    """
+
+    stencil: ['Stencil']
+    stream: ['Stream']
+    equilibrium: ['Equilibrium']
+    collision: ['Collision']
+
+    # optional parameter as there
+    # should be no more than one choice
+    cuda: ['Cuda']
+    lattice: ['Lattice']
+
+    # by default all of these
+    # versions should be created
+    support_no_stream: [bool]
+    support_no_collision: [bool]
+
+    def __init__(self,
+                 stencil: ['Stencil'],
+                 stream: ['Stream'],
+                 equilibrium: ['Equilibrium'],
+                 collision: ['Collision'],
+                 cuda: ['Cuda'] = None,
+                 lattice: ['Lattice'] = None,
+                 support_no_stream: [bool] = None,
+                 support_no_collision: [bool] = None):
+        """
+        :param stencil:
+        :param stream:
+        :param equilibrium:
+        :param collision:
+        :param cuda:
+        :param lattice:
+        :param support_no_stream:
+        :param support_no_collision:
+        """
+
+        self.stencil = stencil
+        self.stream = stream
+        self.equilibrium = equilibrium
+        self.collision = collision
+
+        self.cuda = cuda if cuda is not None else [Cuda()]
+        self.lattice = lattice if lattice is not None else [Lattice()]
+
+        self.support_no_stream = support_no_stream if support_no_stream is not None else [True, False]
+        self.support_no_collision = support_no_collision if support_no_collision is not None else [True, False]
+
+    def entries_(self):
```

Comment:
My thoughts:

1. Building a cuda kernel takes some time. On my computer ~20-30sec.
2. Every option adds a non linear amount of combinations.
3. Pre building all combinations allows the end user to not install all tools needed to compile the extension

4. I am not sure if the code generated is already ready for a JIT compilation. (This needs to be researched)
5. I am not sure if a JIT compiler makes the same optimizations on the code as a "normal" compiler does

At the moment we have:
4x Stencil
1x Stream
1x Equilibrium
1x Collision
optional stream mask
optional collision mask

combinations = 4x1x1x1x2x2 = 16

My local compile time = 16x 30sec = 8min
This is annoying for prototyping! On the other hand it is okay for a one time setup.

If we add more Extensions the compile time will grow polynomial.
This is even more annoying for prototyping. (But not unthinkable for a one time setup)
 
1. First of all it would be interesting to find out if a JIT compiler builds a kernel that is as fast as a pre build kernel.
2. Secondly it would consider to not use a JIT compiler but a "normal" compiler just in time.

E. g.:
- require the end user to have ninja and cuda available.
- Do not build all combinations into one module but every combination into its own module.
- Also do not pre build them but look if the module is available (just in time) and if it is not available generate it and install it.
This way you cache already build modules and the overhead appears only the first time you use a combination after an install.
Problem: You need to version the modules to rebuild them when a new version of lettuce is installed. (or uninstall all combinations when lettuce gets installed/updated)

3. Thirdly it is considerable to pre compile only the most important combinations and JIT compile all others.